File: systems/python-reference-runtime/l1/kernel/ipc.py
"""Kernel IPC — low-level message passing for cross-process sync primitives.

Architecture:
  LockChannel carries LockMessage between processes.
  Each Mutex/Semaphore has a dedicated IPC channel for remote waiters.

  sync.py uses this as optional backend:
    thread-safe (default):   threading.Lock (same-process)
    cross-process (opt-in):  LockChannel over IPC bus (multi-process)
"""

from __future__ import annotations

import logging
import threading
import time
import uuid
from collections import deque
from collections.abc import Callable
from contextlib import suppress
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Any

from .params.kernel import IPC_DEFAULT_PRIORITY, IPC_MSG_ID_LENGTH, IPC_REQUEST_TIMEOUT
from .params.sync import IPC_CHANNEL_PENDING_MAX

logger = logging.getLogger(__name__)


class LockOp(Enum):
    """LockOp — enum of ACQUIRE, RELEASE, STATUS, BOOST."""

    ACQUIRE = auto()
    RELEASE = auto()
    STATUS = auto()
    BOOST = auto()


@dataclass
class LockMessage:
    """LockMessage — lock message record (op, lock_name, agent_id, priority, reply_to)."""

    op: LockOp
    lock_name: str
    agent_id: str = ""
    priority: float = IPC_DEFAULT_PRIORITY
    reply_to: str = ""
    timestamp: float = field(default_factory=time.time)
    msg_id: str = field(default_factory=lambda: uuid.uuid4().hex[:IPC_MSG_ID_LENGTH])


class LockChannel:
    """Dedicated IPC channel for lock operations on a single sync primitive.

    Thread-safe.  Each Mutex / Semaphore can own one LockChannel.
    """

    def __init__(self, name: str, max_pending: int = IPC_CHANNEL_PENDING_MAX):
        self.name = name
        self._max_pending = max_pending
        self._lock = threading.Lock()
        self._queue: deque[LockMessage] = deque()
        self._responses: dict[str, Any] = {}
        self._response_events: dict[str, threading.Event] = {}
        self._handlers: list[Callable[[LockMessage], Any]] = []

    def _trim_backlog(self) -> None:
        """Trim consumed backlog beyond *max_pending* (FIFO).

        Handlers run synchronously inside send(), so messages retained in
        ``_queue`` are historical only — capping them keeps memory bounded.
        ``_responses`` is deliberately NOT trimmed here: request() pops its
        own msg_id on both the success and timeout paths, and evicting here
        would race the delivery order (send/respond pop the waiter event
        before the waiter consumes the reply, so a just-delivered reply
        looks like an orphan and would be silently dropped). Orphans (a
        handler replying after the requester timed out) are bounded — at
        most one per timed-out request.
        """
        while len(self._queue) > self._max_pending:
            self._queue.popleft()

    def send(self, msg: LockMessage) -> str:
        """Enqueue a lock message and notify registered handlers."""
        with self._lock:
            self._queue.append(msg)
            for handler in self._handlers:
                try:
                    reply = handler(msg)
                    if reply is not None:
                        self._responses[msg.msg_id] = reply
                        ev = self._response_events.pop(msg.msg_id, None)
                        if ev:
                            ev.set()
                except Exception as e:
                    logger.error("LockChannel handler error: %s", e)
            self._trim_backlog()
        return msg.msg_id

    def request(self, msg: LockMessage, timeout: float = IPC_REQUEST_TIMEOUT) -> Any:
        """Send a message and block until a response arrives or timeout elapses."""
        event = threading.Event()
        with self._lock:
            self._response_events[msg.msg_id] = event
            self._queue.append(msg)
        event.wait(timeout=timeout)
        # Always clean up our registration so abandoned request events
        # do not accumulate in _response_events (memory leak).
        with self._lock:
            self._response_events.pop(msg.msg_id, None)
            with suppress(ValueError):
                self._queue.remove(msg)  # already trimmed
            self._trim_backlog()
            return self._responses.pop(msg.msg_id, {})

    def respond(self, msg_id: str, data: Any) -> None:
        """Store a response for the given message id and wake the waiting requester."""
        with self._lock:
            self._responses[msg_id] = data
            ev = self._response_events.pop(msg_id, None)
            if ev:
                ev.set()
            self._trim_backlog()

    def register_handler(self, handler: Callable[[LockMessage], Any]) -> None:
        """Append a callback invoked for each message sent on this channel."""
        with self._lock:
            self._handlers.append(handler)

    def pending_count(self) -> int:
        """Return the number of messages currently waiting in the queue."""
        with self._lock:
            return len(self._queue)


class LockBus:
    """Central IPC registry for lock channels — cross-process sync backbone."""

    def __init__(self):
        self._channels: dict[str, LockChannel] = {}
        self._lock = threading.Lock()

    def get_channel(self, name: str) -> LockChannel:
        """Return the named channel, creating it on first access."""
        with self._lock:
            if name not in self._channels:
                self._channels[name] = LockChannel(name)
            return self._channels[name]

    def channel_exists(self, name: str) -> bool:
        """Return True if a channel with the given name has been registered."""
        with self._lock:
            return name in self._channels

    def stats(self) -> dict:
        """Return a mapping of channel name to pending message count."""
        with self._lock:
            return {n: ch.pending_count() for n, ch in self._channels.items()}


_lock_bus: LockBus | None = None
_lock_bus_lock = threading.Lock()


def get_lock_bus() -> LockBus:
    """Get the IPC lock bus singleton."""
    global _lock_bus
    if _lock_bus is None:
        with _lock_bus_lock:
            if _lock_bus is None:
                _lock_bus = LockBus()
    return _lock_bus


def reset_lock_bus() -> None:
    """Reset the lock bus singleton to None (for tests / hot reset)."""
    global _lock_bus
    _lock_bus = None

File: systems/rust-kernel-engine/l1-kernel-rs/tests/core/ipc.rs
//! Independent lock-IPC mechanism tests for the Rust kernel.

use l1_kernel_rs::ipc::{
    IPC_DEFAULT_PRIORITY, IPC_MSG_ID_LENGTH, LockBus, LockChannel, LockMessage, LockOp,
    get_lock_bus, reset_lock_bus,
};
use serde_json::json;
use std::sync::Arc;
use std::thread;
use std::time::Duration;

fn message(op: LockOp, name: &str) -> LockMessage {
    LockMessage::new(op, name).with_agent("agent-1")
}

#[test]
fn message_defaults_and_wire_round_trip_are_stable() {
    let message = message(LockOp::Acquire, "lock:test");
    assert_eq!(message.priority, IPC_DEFAULT_PRIORITY);
    assert_eq!(message.msg_id.len(), IPC_MSG_ID_LENGTH);
    let wire = serde_json::to_value(&message).expect("message json");
    let restored: LockMessage = serde_json::from_value(wire).expect("message decode");
    assert_eq!(restored, message);
    let defaults: LockMessage = serde_json::from_value(json!({
        "op": "ACQUIRE",
        "lock_name": "lock:defaults"
    }))
    .expect("defaults");
    assert_eq!(defaults.priority, IPC_DEFAULT_PRIORITY);
    assert_eq!(defaults.msg_id.len(), IPC_MSG_ID_LENGTH);
}

#[test]
fn send_notifies_handlers_and_stores_reply() {
    let channel = LockChannel::new("lock:test");
    let seen = Arc::new(std::sync::Mutex::new(Vec::new()));
    let seen_handler = Arc::clone(&seen);
    channel.register_handler(move |message| {
        seen_handler
            .lock()
            .expect("seen lock")
            .push(message.lock_name.clone());
        Some(json!({"accepted": true}))
    });
    let message = message(LockOp::Acquire, "lock:test");
    let message_id = channel.send(message.clone());
    assert_eq!(message_id, message.msg_id);
    assert_eq!(*seen.lock().expect("seen lock"), vec!["lock:test"]);
    assert_eq!(
        channel.request(message, Some(Duration::ZERO)),
        json!({"accepted": true})
    );
}

#[test]
fn request_is_woken_by_response_and_timeout_cleans_pending_message() {
    let channel = Arc::new(LockChannel::new("lock:test"));
    let response_channel = Arc::clone(&channel);
    let request_message = message(LockOp::Status, "lock:test");
    let response_id = request_message.msg_id.clone();
    let waiter = thread::spawn(move || {
        response_channel.request(request_message, Some(Duration::from_secs(1)))
    });
    thread::sleep(Duration::from_millis(10));
    channel.respond(&response_id, json!({"granted": true}));
    assert_eq!(waiter.join().expect("waiter"), json!({"granted": true}));
    let timeout_message = message(LockOp::Status, "timeout");
    assert_eq!(
        channel.request(timeout_message, Some(Duration::from_millis(1))),
        json!({})
    );
    assert_eq!(channel.pending_count(), 0);
}

#[test]
fn handler_panics_are_contained_and_backlog_is_bounded() {
    let channel = LockChannel::with_capacity("bounded", 2);
    channel.register_handler(|_| -> Option<serde_json::Value> { panic!("handler failure") });
    channel.send(message(LockOp::Acquire, "bounded"));
    channel.send(message(LockOp::Release, "bounded"));
    channel.send(message(LockOp::Status, "bounded"));
    assert_eq!(channel.pending_count(), 2);
}

#[test]
fn bus_reuses_channels_and_reports_stable_stats() {
    let bus = LockBus::new();
    let first = bus.get_channel("lock:a");
    let second = bus.get_channel("lock:a");
    assert!(Arc::ptr_eq(&first, &second));
    assert!(!bus.channel_exists("lock:b"));
    bus.get_channel("lock:b")
        .send(message(LockOp::Acquire, "lock:b"));
    assert_eq!(bus.stats().get("lock:a"), Some(&0));
    assert_eq!(bus.stats().get("lock:b"), Some(&1));
}

#[test]
fn global_bus_can_be_reset() {
    let first = get_lock_bus();
    let _ = first.get_channel("lock:global");
    reset_lock_bus();
    let second = get_lock_bus();
    assert!(!second.channel_exists("lock:global"));
}
